Spring integration error when using MessageRequestReplyReceiverContext data in observation predicate

**In what version(s) of Spring Integration are you seeing this issue?**

`id 'org.springframework.boot' version '3.1.1'`

`implementation("org.springframework.integration:spring-integration-ip")`

**Describe the bug**

When using the `ObservationPredicate` with Spring boot micrometer, I created some rules in order to observe certain contexts. But when using the context `MessageRequestReplyReceiverContext` and using some of its data in order to create a predicate, the rest of the chain of the application just fails with the exception:
```
java.lang.ClassCastException: class io.micrometer.observation.Observation$Context cannot be cast to class org.springframework.integration.support.management.observation.MessageRequestReplyReceiverContext (io.micrometer.observation.Observation$Context and org.springframework.integration.support.management.observation.MessageRequestReplyReceiverContext are in unnamed module of loader 'app')
```

Our application flow is just a simple TCP nio inbound, with a broker outbound. But it seems to be going wrong before we receive it in the the TCP inbound.


```
        observationRegistry.observationConfig()
                .observationPredicate((name, context) -> shouldBeObserved(context));

```


Example of usage in observation predicate when using headers: 
```
        ...
        
        if (context instanceof MessageRequestReplyReceiverContext messageRequestReplyReceiverContext) {
            return Optional.ofNullable(messageRequestReplyReceiverContext.getCarrier())
                    .map(Message::getHeaders)
                    .map(headers -> !headers.containsKey("SOMEKEY"))
                    .orElse(true);
        }

       ...
```

Example of usage in observation predicate when using payload: 

```
        ...
        
        if (context instanceof MessageRequestReplyReceiverContext messageRequestReplyReceiverContext) {
            return Optional.ofNullable(messageRequestReplyReceiverContext.getCarrier())
                    .map(Message::getPayload)
                    .map(payload -> new String((byte[]) payload, StandardCharsets.UTF_8))
                    .map(payload -> !payload.equals(SOMEDATA))
                    .orElse(true);
        }

       ...
```

Exception:
![image](https://github.com/spring-projects/spring-integration/assets/25689471/74e7cd68-2ca5-490a-9913-f02c81fb398d)

**Expected behavior**

Expected to not break the application when simply casting the payload or using the headers in the generic message.

Right. My theory is correct.
Just modified our `IntegrationObservabilityZipkinTests` for this code:
```
observationRegistry.observationConfig()
					.observationPredicate((name, context) -> !(context instanceof MessageRequestReplyReceiverContext));
```
and got the same exception:
```
Caused by: java.lang.ClassCastException: class io.micrometer.observation.Observation$Context cannot be cast to class org.springframework.integration.support.management.observation.MessageRequestReplyReceiverContext (io.micrometer.observation.Observation$Context and org.springframework.integration.support.management.observation.MessageRequestReplyReceiverContext are in unnamed module of loader 'app')
	at io.micrometer.observation.Observation.observeWithContext(Observation.java:603)
	at org.springframework.integration.gateway.MessagingGatewaySupport.sendAndReceiveWithObservation(MessagingGatewaySupport.java:602)
	at org.springframework.integration.gateway.MessagingGatewaySupport.sendAndReceive(MessagingGatewaySupport.java:553)
	at org.springframework.integration.gateway.MessagingGatewaySupport.sendAndReceiveMessage(MessagingGatewaySupport.java:532)
	at org.springframework.integration.support.management.observation.IntegrationObservabilityZipkinTests$TestMessagingGatewaySupport.process(IntegrationObservabilityZipkinTests.java:179)
```

Not sure what is the fix in Micrometer should be, but that `observeWithContext()` is broken when we are in a `NoopObservation` mode.

--- a/spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java
+++ b/spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java
@@ -595,14 +595,16 @@ public abstract class MessagingGatewaySupport extends AbstractEndpoint
 	private Message<?> sendAndReceiveWithObservation(MessageChannel requestChannel, Object object,
 			Message<?> requestMessage) {
 
+		MessageRequestReplyReceiverContext context =
+				new MessageRequestReplyReceiverContext(requestMessage, getComponentName());
+
 		return IntegrationObservation.GATEWAY.observation(this.observationConvention,
 						DefaultMessageRequestReplyReceiverObservationConvention.INSTANCE,
-						() -> new MessageRequestReplyReceiverContext(requestMessage, getComponentName()),
-						this.observationRegistry)
-				.<MessageRequestReplyReceiverContext, Message<?>>observeWithContext((ctx) -> {
+						() -> context, this.observationRegistry)
+				.observe(() -> {
 					Message<?> replyMessage = doSendAndReceive(requestChannel, object, requestMessage);
 					if (replyMessage != null) {
-						ctx.setResponse(replyMessage);
+						context.setResponse(replyMessage);
 					}
 					return replyMessage;
 				});

--- a/spring-integration-core/src/test/java/org/springframework/integration/support/management/observation/IntegrationObservabilityZipkinTests.java
+++ b/spring-integration-core/src/test/java/org/springframework/integration/support/management/observation/IntegrationObservabilityZipkinTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022 the original author or authors.
+ * Copyright 2022-2023 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -26,12 +26,14 @@ import io.micrometer.tracing.Span;
 import io.micrometer.tracing.test.SampleTestRunner;
 import io.micrometer.tracing.test.simple.SpansAssert;
 
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.context.annotation.AnnotationConfigApplicationContext;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.integration.annotation.EndpointId;
 import org.springframework.integration.annotation.Poller;
 import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.integration.channel.NullChannel;
 import org.springframework.integration.channel.QueueChannel;
 import org.springframework.integration.channel.interceptor.ObservationPropagationChannelInterceptor;
 import org.springframework.integration.config.EnableIntegration;
@@ -64,19 +66,34 @@ public class IntegrationObservabilityZipkinTests extends SampleTestRunner {
 	public SampleTestRunnerConsumer yourCode() {
 		return (bb, meterRegistry) -> {
 			ObservationRegistry observationRegistry = getObservationRegistry();
+
+			observationRegistry.observationConfig()
+					.observationPredicate((name, context) ->
+							!(context instanceof MessageRequestReplyReceiverContext messageRequestReplyReceiverContext)
+									|| !messageRequestReplyReceiverContext.getGatewayName()
+									.equals("skippedObservationInboundGateway"));
+
 			try (AnnotationConfigApplicationContext applicationContext = new AnnotationConfigApplicationContext()) {
 				applicationContext.registerBean(ObservationRegistry.class, () -> observationRegistry);
 				applicationContext.register(ObservationIntegrationTestConfiguration.class);
 				applicationContext.refresh();
 
 				TestMessagingGatewaySupport messagingGateway =
-						applicationContext.getBean(TestMessagingGatewaySupport.class);
+						applicationContext.getBean("testInboundGateway", TestMessagingGatewaySupport.class);
 
 				Message<?> receive = messagingGateway.process(new GenericMessage<>("test data"));
 
 				assertThat(receive).isNotNull()
 						.extracting("payload").isEqualTo("test data");
 				var configuration = applicationContext.getBean(ObservationIntegrationTestConfiguration.class);
+
+				messagingGateway =
+						applicationContext.getBean("skippedObservationInboundGateway",
+								TestMessagingGatewaySupport.class);
+
+				receive = messagingGateway.process(new GenericMessage<>("void data"));
+
+				assertThat(receive).isNull();
 
 				assertThat(configuration.observedHandlerLatch.await(10, TimeUnit.SECONDS)).isTrue();
 			}
@@ -112,7 +129,7 @@ public class IntegrationObservabilityZipkinTests extends SampleTestRunner {
 	@EnableIntegration
 	@EnableIntegrationManagement(
 			observationPatterns = {
-					"${spring.integration.management.observation-patterns:testInboundGateway,queueChannel,observedEndpoint}",
+					"${spring.integration.management.observation-patterns:testInboundGateway,skippedObservationInboundGateway,queueChannel,observedEndpoint}",
 					"${spring.integration.management.observation-patterns:}"
 			})
 	public static class ObservationIntegrationTestConfiguration {
@@ -126,7 +143,7 @@ public class IntegrationObservabilityZipkinTests extends SampleTestRunner {
 		}
 
 		@Bean
-		TestMessagingGatewaySupport testInboundGateway(PollableChannel queueChannel) {
+		TestMessagingGatewaySupport testInboundGateway(@Qualifier("queueChannel") PollableChannel queueChannel) {
 			TestMessagingGatewaySupport messagingGatewaySupport = new TestMessagingGatewaySupport();
 			messagingGatewaySupport.setObservationConvention(
 					new DefaultMessageRequestReplyReceiverObservationConvention() {
@@ -144,6 +161,15 @@ public class IntegrationObservabilityZipkinTests extends SampleTestRunner {
 		@Bean
 		public PollableChannel queueChannel() {
 			return new QueueChannel();
+		}
+
+
+		@Bean
+		TestMessagingGatewaySupport skippedObservationInboundGateway() {
+			TestMessagingGatewaySupport messagingGatewaySupport = new TestMessagingGatewaySupport();
+			messagingGatewaySupport.setRequestChannel(new NullChannel());
+			messagingGatewaySupport.setReplyTimeout(0);
+			return messagingGatewaySupport;
 		}
 
 		@Bean

